# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/webrtc/libwebrtcglue/WebrtcVideoCodecFactory.h
# Commit: 709078257c62
# Full Hash: 709078257c628fae054382894de5864b1f5f5c2a
# Author: Andreas Pehrson <apehrson@mozilla.com>
# Date: 2021-11-01 18:17:33
# Regressor Bug: 1654112
# File Overlap Count: 3
# Description:
#   Bug 1654112 - Re-wire GMP PluginID between codecs and VideoConduit. r=bwc
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D104724
# ==============================================================================

diff -r 3ca24e761cc3 -r 709078257c62 dom/media/webrtc/libwebrtcglue/WebrtcVideoCodecFactory.h
--- a/dom/media/webrtc/libwebrtcglue/WebrtcVideoCodecFactory.h	Wed Feb 10 10:34:14 2021 +0100
+++ b/dom/media/webrtc/libwebrtcglue/WebrtcVideoCodecFactory.h	Wed Feb 10 15:17:29 2021 +0100
@@ -11,7 +11,41 @@
 #include "MediaEventSource.h"
 
 namespace mozilla {
-class WebrtcVideoDecoderFactory : public webrtc::VideoDecoderFactory {
+class GmpPluginNotifier {
+ public:
+  GmpPluginNotifier()
+      : mOwningThread(GetCurrentSerialEventTarget()),
+        mCreatedGmpPluginEvent(mOwningThread),
+        mReleasedGmpPluginEvent(mOwningThread) {}
+
+  ~GmpPluginNotifier() {
+    mOwningThread->Dispatch(NS_NewRunnableFunction(
+        "~GmpPluginNotifier",
+        [createdEvent = std::move(mCreatedGmpPluginEvent),
+         releasedEvent = std::move(mReleasedGmpPluginEvent)]() mutable {
+          createdEvent.DisconnectAll();
+          releasedEvent.DisconnectAll();
+        }));
+  }
+
+  MediaEventSource<uint64_t>& CreatedGmpPluginEvent() {
+    MOZ_ASSERT(mOwningThread->IsOnCurrentThread());
+    return mCreatedGmpPluginEvent;
+  }
+
+  MediaEventSource<uint64_t>& ReleasedGmpPluginEvent() {
+    MOZ_ASSERT(mOwningThread->IsOnCurrentThread());
+    return mReleasedGmpPluginEvent;
+  }
+
+ protected:
+  const nsCOMPtr<nsISerialEventTarget> mOwningThread;
+  MediaEventForwarder<uint64_t> mCreatedGmpPluginEvent;
+  MediaEventForwarder<uint64_t> mReleasedGmpPluginEvent;
+};
+
+class WebrtcVideoDecoderFactory : public GmpPluginNotifier,
+                                  public webrtc::VideoDecoderFactory {
  public:
   explicit WebrtcVideoDecoderFactory(std::string aPCHandle)
       : mPCHandle(std::move(aPCHandle)) {}
@@ -28,8 +62,10 @@
   const std::string mPCHandle;
 };
 
-class WebrtcVideoEncoderFactory : public webrtc::VideoEncoderFactory {
-  class InternalFactory : public webrtc::VideoEncoderFactory {
+class WebrtcVideoEncoderFactory : public GmpPluginNotifier,
+                                  public webrtc::VideoEncoderFactory {
+  class InternalFactory : public GmpPluginNotifier,
+                          public webrtc::VideoEncoderFactory {
    public:
     explicit InternalFactory(std::string aPCHandle)
         : mPCHandle(std::move(aPCHandle)) {}
@@ -50,7 +86,10 @@
 
  public:
   explicit WebrtcVideoEncoderFactory(std::string aPCHandle)
-      : mInternalFactory(MakeUnique<InternalFactory>(std::move(aPCHandle))) {}
+      : mInternalFactory(MakeUnique<InternalFactory>(std::move(aPCHandle))) {
+    mCreatedGmpPluginEvent.Forward(mInternalFactory->CreatedGmpPluginEvent());
+    mReleasedGmpPluginEvent.Forward(mInternalFactory->ReleasedGmpPluginEvent());
+  }
 
   std::vector<webrtc::SdpVideoFormat> GetSupportedFormats() const override {
     MOZ_CRASH("Unexpected call");
